### DISCUSSION TITLE
Having Issue replacing array in React Hook Forms v7.28.1

### DISCUSSION THREAD

**Discussion by BologneseBandit** [react-hook-form/react-hook-form#8067] [Q&A] [answered]
I am making a form component that requires you to denote the number of pieces you want to make to split a circle. Based on the number of pieces you can enter the angles for each of the parts sucht that the total adds to 360. I have completed the validation, and that part works smoothly, however when I unmount the form component (storing the form state in a parent component) and remount the form component, I was hoping to use useFieldArray's replace to restore the previous values of the split. 

I am having an issue as the field arrays are not being resorted to their previous state. I will quote relavent parts of the code below, as it is not a public repo.

### Setting up the UseForm & useFieldArray objects
Note: Offset angle is another independent variable of the field array
```
const {
        register: registerSeg,
        watch: watchSeg,
        reset: resetSeg,
        control
    } = useForm({
        mode: "onChange",
        defaultValues: {
            offset_angle: segmentationData.offset_angle,
        }
    })

  const {fields, append, remove, replace } = useFieldArray({
    control,
    name: 'segment_angles',
    shouldUnregister: false
  })
```

### OnChange function for input denoting number of parts
```
onChange={(e) => {
  const targetSegments = parseInt(e.target.value)
  if(targetSegments > fields.length){
    for( let i = fields.length; i < targetSegments; i++ ){
      append({angle: 0}, {shouldFocus: false})
    }
  }
   else if(targetSegments < fields.length) {
     for(let i = fields.length - 1; i > targetSegments - 1; i--){
       remove(i)
     }
   }
  //This is updating the component state of the number of segments.
  setNumberOfSegments(targetSegments) 
}}
```
Note: On mount setNumberOfComponents is set to either 0 or if segmentationData.segment_angles has a length value, it is set to the length of the object.

### Side effect to set up fields on mount of the form component
```
useEffect(() => {
        //Replace filed array with current fields if remounted
        if(segmentationData.segment_angles !== undefined) {
            replace(segmentationData.segment_angles)
        }
    },[])
```
For content, when logging segmentation data before replace, it contains and array of object with the key angle, as an example below.
[{angle:180},{angle:180}] for a 2 part split with 180degrees for each piece.
After this replace fields is equal to an empty array.

### Mapping of Fields to the DOM
```
fields.map(({ id },index) => {
  return (
   <div className="input-group mb-3" key={`angle_segment_${index}`}>
      <span className="input-group-text">{`Segment #${index + 1}`}</span>
      <input className="form-control" type="number" name={`segment_angles.${index}.angle`} placeholder="0" min="0" max="359" {... registerSeg(`segment_angles.${index}.angle`)} />
  </div>
  )
})
```

Would apprciate any help, or tips to debug as I am not sure what I am missing. Also if there are other best practise that I have overlooked please let me know.

**Comment by Moshyfawn** [comment, marked as answer]
I tried to quickly reproduce your use-case in [this CSB](https://codesandbox.io/s/hardcore-stonebraker-pup4si?file=/src/App.tsx) and got it working. Can you take a look and try tweaking it to closer match your use-case?

**Reply by BologneseBandit** [reply]
Whilst reproducing I found the issue. It seems that it was being set, but I had an effect hook that was running after the empty array dependancy (only on remount) which was calling replace({}) on the field array when parent triggered all forms to be cleared.

So there was no error relating to RHFs, just an error with how I had arranged my code. Thanks for asking me to tweak the CSB as that was the eureka moment. Also here is the [tweaked CSB](https://codesandbox.io/s/gifted-wilson-di8b6g) if you were wondering about the setup.

**Reply by Moshyfawn** [reply]
No problem!
